st2auth_ldap_backend/ldap_backend.py

```diff
@@ -160,15 +164,52 @@ def _ldap_search(self, connection, username, criteria):
                          (base_dn, search_filter, scope, pattern)
         :type criteria: ``dict``
         """
+        results = []
         base_dn = criteria['base_dn']
         search_filter = criteria['search_filter'].format(username=username)
         scope = self._scope_to_ldap_option(criteria['scope'])
 
         LOG.debug('Searching ... %s %s %s' % (base_dn, scope, search_filter))
-        result = connection.search_s(base_dn, scope, search_filter)
+
+        result_id = connection.search(base_dn, scope, search_filter)
+        while 1:
+            r_type, r_data = connection.result(result_id, all=0)
+
+            if r_type == ldap.RES_SEARCH_RESULT:
+                break
+            elif r_type == ldap.RES_SEARCH_ENTRY:
+                results.append(r_data[0])
+            elif r_type == ldap.RES_SEARCH_REFERENCE:
+                # extract referral informations (uri and bind_dn)
```

Comment:
This `elif` content seems prime for breaking out into a private method for the class. This will help readability and help developers understand what's going on better.